src/js/components/Pagination/Pagination.js

```diff
@@ -0,0 +1,202 @@
+      numMiddlePages: numMiddlePagesProp = 3,
+      onChange,
+      page: pageProp,
+      step = 10,
+      ...rest
+    },
+    ref,
+  ) => {
+    const theme = useContext(ThemeContext) || defaultProps.theme;
+
+    const [activePage, setActivePage] = useState(pageProp || 1);
+
+    /* Calculate total number pages */
+    const totalPages = Math.ceil(numItems / step);
+
+    /* Define page indices to display */
+    const beginPages = getPageIndices(1, Math.min(numEdgePages, totalPages));
+    const endPages = getPageIndices(
+      Math.max(totalPages - numEdgePages + 1, numEdgePages + 1),
+      totalPages,
+    );
+
+    let numMiddlePages;
+    if (numMiddlePagesProp < 1) {
+      numMiddlePages = 1;
+      console.warn(
+        // eslint-disable-next-line max-len
+        `Property "numMiddlePages" should not be < 1. One middle page button will be shown. Set "numMiddlePages" >= 1 to remove this warning.`,
+      );
+    } else numMiddlePages = numMiddlePagesProp;
+
+    let startingMiddlePages;
+    // odd
+    if (numMiddlePages % 2)
+      startingMiddlePages = Math.min(
+        activePage - Math.floor(numMiddlePages / 2),
+        totalPages - numEdgePages - numMiddlePages,
+      );
+    // even, cannot split equally around active page
+    // let extra page appear on middlePagesEnd instead
+    else
+      startingMiddlePages = Math.min(
+        activePage - Math.floor(numMiddlePages / 2) + 1,
+        totalPages - numEdgePages - numMiddlePages,
+      );
+
+    const middlePagesBegin = Math.max(startingMiddlePages, numEdgePages + 2);
+    const middlePagesEnd = Math.min(
+      Math.max(
+        activePage + Math.floor(numMiddlePages / 2),
+        numEdgePages + numMiddlePages + 1,
+      ),
+      endPages.length > 0 ? endPages[0] - 2 : totalPages - 1,
+    );
+
+    const middlePages = getPageIndices(middlePagesBegin, middlePagesEnd);
+
+    let beginFlex = [];
+    if (middlePagesBegin > numEdgePages + 2) beginFlex = ['more-prev'];
+    else if (numEdgePages + 1 < totalPages - numEdgePages)
+      beginFlex = [numEdgePages + 1];
+
+    let endFlex = [];
+    if (middlePagesEnd < totalPages - numEdgePages - 1) endFlex = ['more-next'];
+    else if (totalPages - numEdgePages > numEdgePages)
+      endFlex = [totalPages - numEdgePages];
+
+    const getItemIndices = nextPage => {
+      const startIndex = step * (nextPage - 1);
+      const endIndex = startIndex + step;
+      return { startIndex, endIndex };
+    };
+
+    const handleClick = (event, nextPage) => {
+      event.persist();
+      const adjustedEvent = event;
+      adjustedEvent.page = nextPage;
+
+      // for controlled use cases, provide user with info on
+      // what range of indices should be displayed given the active page
+      const { startIndex, endIndex } = getItemIndices(nextPage);
+      adjustedEvent.startIndex = startIndex;
+      adjustedEvent.endIndex = endIndex;
+
+      setActivePage(event.page);
+
+      if (onChange) {
```

Comment:
I think we can put the `event.persist()` and almost everything else within this `if (onChange)` and just call `setActivePage(nextPage)` otherwise.

Reply:
To confirm, if user provides onChange then we won't increment the page for them?

Reply:
or we increment the page, and then apply whatever additional onChange functionality they provide.

Reply:
> or we increment the page, and then apply whatever additional onChange functionality they provide.

this one ^^^